pkg/liqo-controller-manager/webhooks/shadowpod/cache.go

```diff
@@ -0,0 +1,262 @@
+// Copyright 2019-2022 The Liqo Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shadowpod
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	sharing "github.com/liqotech/liqo/apis/sharing/v1alpha1"
+	vkv1alpha1 "github.com/liqotech/liqo/apis/virtualkubelet/v1alpha1"
+	"github.com/liqotech/liqo/pkg/discovery"
+	fcutils "github.com/liqotech/liqo/pkg/utils/foreignCluster"
+)
+
+type peeringCache struct {
+	peeringInfo sync.Map
+	ready       bool
+}
+
+/**
+ * PeeringCache methods
+ */
+
+// CacheRefresher is a wrapper funcion that receives a ShadowPodValidator and starts an PollImmediateInfinite timer to periodically refresh the cache.
+func CacheRefresher(spv *Validator, interval time.Duration) func(ctx context.Context) error {
+	return func(ctx context.Context) error {
+		return wait.PollImmediateInfiniteWithContext(ctx, interval, spv.refreshCache)
+	}
+}
+
+func (spv *Validator) initializeCache(ctx context.Context) (err error) {
+	resourceOfferList := sharing.ResourceOfferList{}
+	if err := spv.client.List(ctx, &resourceOfferList, &client.ListOptions{}); err != nil {
+		return err
+	}
+	cacherefreshlog.Info("Cache initialization started")
+	for i := range resourceOfferList.Items {
+		ro := &resourceOfferList.Items[i]
+		clusterID := ro.Labels[discovery.ClusterIDLabel]
+		clusterName := retrieveClusterName(ctx, spv.client, clusterID)
+		if clusterID == "" {
+			return fmt.Errorf("ResourceOffer %s has no cluster id", clusterName)
+		}
```

Comment:
Should probably avoid blocking completely and just skip this entry, to prevent a malformed request from breaking the creation of shadow pods in the entire cluster.